ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,777 @@
+      while (inFlightRequestsIterator.hasNext()) {
+        Map.Entry<Integer, GetRequestInfo> entry = inFlightRequestsIterator.next();
+        if (time.milliseconds() - entry.getValue().startTimeMs > routerConfig.routerRequestTimeoutMs) {
+          chunkOperationTracker.onResponse(entry.getValue().replicaId, false);
+          chunkException = new RouterException("Timed out waiting for responses", RouterErrorCode.OperationTimedOut);
+          inFlightRequestsIterator.remove();
+        } else {
+          // the entries are ordered by correlation id and time. Break on the first request that has not timed out.
+          break;
+        }
+      }
+    }
+
+    /**
+     * Fetch {@link GetRequest}s to send for the current data chunk.
+     */
+    private void fetchRequests(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
+      Iterator<ReplicaId> replicaIterator = chunkOperationTracker.getReplicaIterator();
+      while (replicaIterator.hasNext()) {
+        ReplicaId replicaId = replicaIterator.next();
+        replicaIterator.remove();
+        String hostname = replicaId.getDataNodeId().getHostname();
+        Port port = replicaId.getDataNodeId().getPortToConnectTo();
+        GetRequest getRequest = createGetRequest(chunkBlobId, getOperationFlag());
+        RequestInfo request = new RequestInfo(hostname, port, getRequest);
+        int correlationId = getRequest.getCorrelationId();
+        correlationIdToGetRequestInfo.put(correlationId, new GetRequestInfo(replicaId, time.milliseconds()));
+        correlationIdToGetChunk.put(correlationId, this);
+        requestRegistrationCallback.registerRequestToSend(GetBlobOperation.this, request);
+        state = ChunkState.InProgress;
+      }
+    }
+
+    /**
+     * Check if the operation on the chunk is eligible for completion, if so complete it.
+     */
+    void checkAndMaybeComplete() {
+      if (chunkOperationTracker.isDone()) {
+        if (chunkOperationTracker.hasSucceeded()) {
+          // override any previously set exceptions.
+          chunkException = null;
+        }
+        chunkCompleted = true;
+      }
+      if (chunkCompleted) {
+        operationException.set(chunkException);
+        state = ChunkState.Complete;
+      }
+    }
+
+    /**
+     * Handle the body of the response: Deserialize and add to the list of chunk buffers.
+     * @param payload the body of the response.
+     * @throws IOException if there is an IOException while deserializing the body.
+     * @throws MessageFormatException if there is a MessageFormatException while deserializing the body.
+     */
+    void handleBody(InputStream payload)
+        throws IOException, MessageFormatException {
+      if (!successfullyDeserialized) {
+        BlobData blobData = MessageFormatRecord.deserializeBlob(payload);
+        chunkIndexToBuffer.put(chunkIndex,
+            ByteBuffer.wrap(Utils.readBytesFromStream(blobData.getStream(), (int) blobData.getSize())));
```

Comment:
1. Will the payload get duplicated? In `blobData` the payload is materialized in memory as `ByteBufferInputStream`, and `Utils.readBytesFromStream` create a byte array, which double the memory usage. Would this be necessary?

2. If `BlobData` does not always have its `stream` as a `ByteBuffer`, this method would be blocking?